File: Minimum_Cost_Tree_from_leav_values.py
# -*- coding: utf-8 -*-

#####################################################################
# File Name:  Minimum_Cost_Tree_from_leav_values.py
# Author: shenming
# Created Time: Sun Jan 26 20:04:33 2020
# reference: https://leetcode.com/problems/minimum-cost-tree-from-leaf-values/discuss/478708/RZ-Summary-of-all-the-solutions-I-have-learned-from-Discuss-in-Python
#####################################################################

import os
import sys

# Top down code with memorization
# class Solution_ref_1:
#     def mctFromLeafValues(self, arr: list[int]) -> int:
#         return self.helper(arr, 0, len(arr) - 1, {})
#
#     def helper(self, arr, l, r, cache):
#         if (l, r) in cache:
#             return cache[(l, r)]
#         if l>= r:
#             return 0
#         res = float('inf')
#         for i in range(l, r):
#             rootVal = max(arr[l:i+1]) * max(arr[i+1:r+1])
#             res = min(res, rootVal + self.helper(arr, l, i, cache) + self.helper(arr, i+1, r, cache))
#         cache[(l, r)] = res
#         return res


class Solution:
    def mctFromLeafValues(self, arr) -> int:
        stack = [float('inf')]
        res = 0
        for num in arr:
            while stack and stack[-1] <= num:
                cur = stack.pop()
                if stack:
                    res += cur * min(stack[-1], num)
            stack.append(num)
        while len(stack) > 2:
            res += stack.pop() * stack[-1]
        return res

if __name__ == "__main__":
    s = Solution()
    print(s.mctFromLeafValues([6, 3, 7, 1, 2]))

File: Minimum_Cost_Tree_from_leav_values.cpp
/*************************************************************************
> File Name:  Minimum_Cost_Tree_from_leav_values.cpp
> Author: shenming
> Created Time: Sun Jan 26 19:22:52 2020
************************************************************************/

#include "common_lc.h"
class Solution {
public:
    map<pair<int, int>, pair<long long, int>> mem;

    pair<long long, int> dp(int be, int en, vector<int>& arr) {
        if (en - be == 1) { // leaf
            return pair<long long, int>(0, arr[be]);
        }
        pair<int, int> key(be, en);
        if (mem.count(key)) {
            return mem[key];
        }
        int second = -1;
        for (int i = be; i < en; i++) {
            if (arr[i] > second) {
                second = arr[i];
            }
        }
        long long mim_first = LONG_MAX;
        for (int i = be + 1; i < en; i++) {
            pair<long long, int> ll = dp(be, i, arr);
            pair<long long, int> rr = dp(i, en, arr);
            long long first = ll.first + rr.first + ll.second * rr.second;
            if (first < mim_first) {
                mim_first = first;
            }
        }
        mem[pair<int, int>(be, en)] = pair<long long, int>(mim_first, second);
        return pair<long long, int>(mim_first, second);
    }

    int mctFromLeafValues(vector<int>& arr) {
        vector<pair<int, int>> pair_arr;
        pair<long long, int> ret = dp(0, arr.size(), arr);
        return (int)ret.first;
    }
};

int main()
{
    Solution s;
    vector<int> nums = {6, 2, 4};
    int ret = s.mctFromLeafValues(nums);
    printf("%d\n", ret);
    return 0;
}
